craft_parts/plugins/npm_plugin.py

```diff
@@ -0,0 +1,182 @@
+
+class NpmPluginEnvironmentValidator(validator.PluginEnvironmentValidator):
+    """Check the execution environment for the npm plugin.
+
+    :param part_name: The part whose build environment is being validated.
+    :param env: A string containing the build step environment setup.
+    """
+
+    def validate_environment(self, *, part_dependencies: Optional[List[str]] = None):
+        """Ensure the environment has the dependencies to build npm applications.
+
+        :param part_dependencies: A list of the parts this part depends on.
+        """
+        for dependency in ["node", "npm"]:
+            self.validate_dependency(
+                dependency=dependency, part_dependencies=part_dependencies
+            )
+
+
+class NpmPlugin(Plugin):
+    """A plugin for npm projects.
+
+    This plugin uses the common plugin keywords as well as those for "sources".
+    For more information check the 'plugins' topic for the former and the
+    'sources' topic for the latter.
+
+    Additionally, this plugin uses the following plugin-specific keywords:
+        - npm-include-node
+          (bool; default: False)
+          If true, download and include the node binary and its dependencies.
+          If npm-include-node is true, then npm-node-version must be defined.
+        - npm-node-version
+          (str: default: None)
+          Which version of node to download (e.g. "16.14.2")
+    """
+
+    properties_class = NpmPluginProperties
+    validator_class = NpmPluginEnvironmentValidator
+
+    @staticmethod
+    def _get_architecture() -> str:
+        """Get system architecture, formatted for downloading node.
+
+        :raise InvalidArchitecture: If the system architecture
+        isn't compatible with node.
+        """
+        snap_arch = os.getenv("SNAP_ARCH")
+        if snap_arch is not None:
+            try:
+                node_arch = _NODE_ARCH_FROM_SNAP_ARCH[snap_arch]
+            except KeyError as error:
+                raise InvalidArchitecture(arch_name=snap_arch) from error
+        else:
+            machine_type = platform.machine()
+            architecture_type = platform.architecture()
+            try:
+                node_arch = _NODE_ARCH_FROM_PLATFORM[machine_type][architecture_type[0]]
+            except KeyError as error:
+                raise InvalidArchitecture(
+                    arch_name=f"{machine_type} {architecture_type}"
+                ) from error
+
+        return node_arch
+
+    def get_build_snaps(self) -> Set[str]:
+        """Return a set of required snaps to install in the build environment."""
+        return set()
+
+    def get_build_packages(self) -> Set[str]:
+        """Return a set of required packages to install in the build environment."""
+        return {"curl", "gcc"}
+
+    def get_build_environment(self) -> Dict[str, str]:
+        """Return a dictionary with the environment to use in the build step."""
+        return dict(PATH="${CRAFT_PART_INSTALL}/bin:${PATH}")
```

Comment:
this should only be set if options.npm_include_node is True, right?

Reply:
Right. And the same is true for `curl` for `get_build_packages()`.  
I added logic and unit tests for both.